aten/src/TH/generic/THTensorRandom.cpp

```diff
@@ -477,32 +481,81 @@ void THTensor_(getRNGState)(at::Generator *_generator, THTensor *self)
 {
   // See Note [Thread-safety and Generators]
   std::lock_guard<std::mutex> lock(_generator->mutex_);
-  static const size_t size = sizeof(at::CPUGenerator);
-  at::CPUGenerator* rng_state;
+  static const size_t size = sizeof(THGeneratorStateNew);
   THTensor_(resize1d)(self, size);
   THArgCheck(THTensor_(nElement)(self) == size, 1, "RNG state is wrong size");
   THArgCheck(THTensor_(isContiguous)(self), 1, "RNG state needs to be contiguous");
-  rng_state = (at::CPUGenerator*)self->data<scalar_t>();
+
+  // cast byte tensor to POD type
+  THGeneratorStateNew* rng_state = (THGeneratorStateNew*)self->data<scalar_t>();
+
+  // accumulate generator data to be copied into byte tensor
+  auto state = c10::guts::make_unique<THGeneratorStateNew>();
   auto cast_generator = static_cast<at::CPUGenerator*>(_generator);
-  memcpy(rng_state, cast_generator, size);
+  state->engine = cast_generator->engine();
+  state->next_float_normal_sample = cast_generator->next_float_normal_sample();
+  state->next_double_normal_sample = cast_generator->next_double_normal_sample();
+
+  memcpy(rng_state, state.get(), size);
 }
 
 void THTensor_(setRNGState)(at::Generator *_generator, THTensor *self)
 {
   // See Note [Thread-safety and Generators]
   std::lock_guard<std::mutex> lock(_generator->mutex_);
-  static const size_t size = sizeof(at::CPUGenerator);
-  at::CPUGenerator* rng_state;
-  THArgCheck(THTensor_(nElement)(self) == size, 1, "RNG state is wrong size");
+  auto cast_generator = static_cast<at::CPUGenerator*>(_generator);
```

Comment:
I didn't catch this previously, but this now looks a little skeevy to me. Can you guarantee that we actually have a `CPUGenerator` at this point in time? If we can't guarantee it, you better have a check.